compio-driver/src/buffer_pool/fallback.rs

```diff
@@ -0,0 +1,221 @@
+//! The fallback buffer pool. It is backed by a [`VecDeque`] of [`Vec<u8>`].
+//! An [`OwnedBuffer`] is selected when the op is created. It keeps a strong
+//! reference to the buffer pool. The [`BorrowedBuffer`] is created after the op
+//! returns successfully.
+
+use std::{
+    borrow::{Borrow, BorrowMut},
+    cell::RefCell,
+    collections::VecDeque,
+    fmt::{Debug, Formatter},
+    io,
+    mem::ManuallyDrop,
+    ops::{Deref, DerefMut},
+    rc::Rc,
+};
+
+use compio_buf::{IntoInner, IoBuf, IoBufMut, SetBufInit, Slice};
+
+struct BufferPoolInner {
+    buffers: RefCell<VecDeque<Vec<u8>>>,
+}
+
+impl BufferPoolInner {
+    pub(crate) fn add_buffer(&self, mut buffer: Vec<u8>) {
+        buffer.clear();
+        self.buffers.borrow_mut().push_back(buffer)
+    }
+}
+
+/// Buffer pool
+///
+/// A buffer pool to allow user no need to specify a specific buffer to do the
+/// IO operation
+pub struct BufferPool {
+    inner: Rc<BufferPoolInner>,
+}
+
+impl Debug for BufferPool {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("BufferPool").finish_non_exhaustive()
+    }
+}
+
+impl BufferPool {
+    pub(crate) fn new(buffer_len: u16, buffer_size: usize) -> Self {
+        // To match the behavior of io-uring, extend the number of buffers.
+        let buffers = (0..buffer_len.next_power_of_two())
+            .map(|_| Vec::with_capacity(buffer_size))
+            .collect();
+
+        Self {
+            inner: Rc::new(BufferPoolInner {
+                buffers: RefCell::new(buffers),
+            }),
+        }
+    }
+
+    /// Select an [`OwnedBuffer`] when the op creates.
+    pub(crate) fn get_buffer(&self, len: usize) -> io::Result<OwnedBuffer> {
```

Comment:
Why do we have `io::Result` for fallback while `Option` for iour?

Reply:
Good catch! Fixed.